docs/source/format/Columnar.rst

```diff
@@ -765,6 +765,72 @@ application.
 We discuss dictionary encoding as it relates to serialization further
 below.
 
+.. _run-end-encoded-layout:
+
+Run-End Encoded Layout
+-------------------------
+
+Run-end encoding (REE) is a variation of run-length encoding (RLE). These
+encodings are well-suited for representing data containing sequences of the
+same value, called runs. In run-end encoding, each run is represented as a
+value and an integer giving the index in the array where the run ends.
+
+Any array can be run-end encoded. A run-end encoded array has no buffers
+by itself, but has two child arrays. The first one holds a signed integer
+called a "run end" for each run. The run ends array can hold either 16, 32, or
+64-bit integers. The actual values of each run are held in the second child array.
+For the purposes of determining field names and schemas, these child arrays
+are prescribed the standard names of **run_ends** and **values** respectively.
+
+The values in the first child array represent the length of each run. They do
+not hold the length of the respective run directly, but the accumulated length
+of all runs from the first to the current one, i.e. the logical index where the
+current run ends. This allows relatively efficient random access from a logical
+index using binary search. The length of an individual run can be determined by
+subtracting two adjacent values. (Contrast this with run-length encoding, in
+which the lengths of the runs are represented directly, and in which random
+access is less efficient.)
+
+A run must have have a length of at least 1. This means the values in the
+run ends array all positive and in strictly ascending order. A run end cannot be
+null.
+
+As an example, you could have the following data: ::
+
+    type: Float32
+    [1.0, 1.0, 1.0, 1.0, null, null, 2.0]
+
+In Run-end-encoded form, this could appear as:
+
+::
+
+    * Length: 7, Null count: 2
+    * Children arrays:
+
+      * run_ends (Int32):
+        * Length: 3, Null count: 0
+        * Validity bitmap buffer: Not required
```

Comment:
The run ends can have nulls?

Reply:
there shouldn't be any nulls in the run_ends array, i'll make that explicit. Good catch

Reply:
Actually it's already stated just above these examples that a run end cannot be null, i'll just update this diagram to be explicit.

Reply:
That might have been discussed before(and probably also part of the vote), but if they can't have nulls and thus basically are a single buffer, is there a reason that run_ends are not a "buffer" of the array instead of the first "child array" ? (like variable size list has offsets as buffer and values as child array)

Reply:
(have them as a buffer would avoid the ambiguity around the validity bitmap of the run_ends as child array)

Reply:
found the mailing list discussion on this which explicitly addresses being children arrays rather than a buffer: https://lists.apache.org/thread/bfz3m5nyf7flq7n6q9b1bx3jhcn4wq29 

TL;DR: in addition to the point I made above, there's this comment from @westonpace at the end of the thread:

> * There are two child arrays not necessarily for slicing but more to help
distinguish between the logical length (there are no buffers with the
logical length but there is an array) and the physical length (all of the
buffers are in these child arrays which are sized to the physical length).

Reply:
It might be good to preserve that reasoning somewhere more easily findable for posterity?

Reply:
added to `Columnar.rst`. Let me know if that's sufficient or if anyone thinks I should modify it a bit.